File: SearchInSortedArray.java
class SearchInSortedArray {
    public int search(int[] nums, int target) {
        int start = 0;
        int end = nums.length-1;
        while(start <= end){
            int mid = (start + end)/2;
            if (nums[mid] == target) return mid;
            if (nums[mid] >= nums[start]) {
                //means left half of the array is sorted
                if (nums[mid]>=target && nums[start]<=target)
                    end = mid-1;  //if the target lies in the left half, bring the end to the left half
                else start = mid+1; //otherwise, bring start to the right half and search in right half of the array
            }
            else{
                //means right half of the array is sorted
                if(nums[mid]<=target && nums[end]>=target)
                    start = mid+1; //if the target lies in the right half, bring the start to the right half
                else end=mid-1; //otherwise, bring end to the left half and search in left half of the array
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        SearchInSortedArray obj = new SearchInSortedArray();
        int[] nums = new int[]{4,5,6,7,0,1,2};
        int target = 4;
        int res = obj.search(nums, target);
        System.out.println("result = " + res);

    }
}

File: search_in_sorted_array.cpp
// Time Complexity : O(logn)
// Space Complexity : O(logn) //stack space
// Any problem you faced while coding this :
// Your code here along with comments explaining your approach
//1. Find the sorted region and whether the target is present in the sorted region.
//2. If the target is not present explore unsorted region in both the cases.
//3. If the target is not found return -1.

class Solution {
public:
    int search(vector<int>& nums, int target) {
        //edge
        if(nums.size()==0){
            return -1;
        }
        //recurse
        int size = nums.size();
        return bin_search(nums, 0, size-1, size, target);
    }
    
    int bin_search(vector<int> & nums, int low, int high, int& size,int target){
     int result =-1;
        if(low<=high){
            int mid = low +(high-low)/2;
            //perfect match
            if(nums[mid] == target){
                return mid;
            }
            
            //sorted region
            if(nums[mid] >= nums[low] && target>=nums[low] && target<nums[mid]){
               result =  bin_search(nums, low, mid-1, size, target);
                return result;
            }else if(nums[mid] < nums[high] && target>nums[mid] && target<=nums[high]){ 
                 result = bin_search(nums, mid+1, high, size, target);
                return result;
            }else{ //unsorted region
                result = bin_search(nums,mid+1, high, size, target);
                if(result==-1)
                    result =bin_search(nums, low, mid-1, size, target);
                return result;
            }
        }
        
       return result;
    }
};
